### DISCUSSION TITLE
Expanding and Selecting in TreeModelImpl

### DISCUSSION THREAD

**Discussion by maheshrajrp** [eclipse-theia/theia#10676] [Q&A] [answered]
Hi,
I'm using `CustomTreeModel` that extends `TreeModelImp`. I need help with scenario.
Consider this Tree,

![image](https://user-images.githubusercontent.com/80001447/151313244-8775d5e5-4e60-45e2-8c69-16ee04c04e85.png)

Now, assume all the nodes are in collapsed state and only root is visible. If I need to select the child2 and make parent1 expanded along the way. How should I move forward using just the node data.

I tried variations such as `this.selectNode(node)` but didn't work as the parent nodes aren't expanded and so the selection isn't visible. As I'm unsure if there is a bug in the `CustomTreeModel` and I need to confirm if I am doing the right way.  

Thanks

**Comment by colin-grant-work** [comment, marked as answer]
Hi, @maheshrajrp, thanks for your question. In your scenario, it should be possible to do something like this

1. Get a reference to the child (e.g. `getNode(child.id)`)
2. Traverse the tree and expand each parent:
```typescript
let toExpand = myChild;
while (toExpand = toExpand.parent) {
    await this.expandNode(toExpand);
}
```
An example of where something like this is done is in the File tree, though that case is a bit more complex:

https://github.com/eclipse-theia/theia/blob/541b300adc029ab1dd729da1ca49179ace1447b2/packages/navigator/src/browser/navigator-model.ts#L171-L214

**Reply by maheshrajrp** [reply]
Hi @colin-grant-work ,
Thanks !! This works perfect.
